specs/AppLifecycle/Restart/restartApi.markdown

```diff
@@ -0,0 +1,259 @@
+</li>
+<li>
+<p>2: InvalidUser - could not restart for the specified user.</p>
+</li>
+<li>
+<p>3: Other - unspecified failure.</p>
+</li>
+</ul></td>
+</tr>
+<tr class="odd">
+<td>class <strong>LaunchActivatedEventArg</strong></td>
+<td><p>Relevant Properties:</p>
+<ul>
+<li><p>Arguments: gets the arguments that are passed to the app during its launch activation.</p></li>
+<li><p>Kind: Gets the reason that this app is being activated. This is of type enum ActivationKind</p></li>
+</ul>
+<p>PreviousExecutionState: Gets the execution state of the app before this activation.</p></td>
+</tr>
+</tbody>
+</table>
+
+## New API Details
+
+### RequestRestartNow
+
+<table>
+<colgroup>
+<col style="width: 41%" />
+<col style="width: 53%" />
+<col style="width: 5%" />
+</colgroup>
+<thead>
+<tr class="header">
+<th>New API </th>
+<th>Description</th>
+</tr>
+</thead>
+<tbody>
+<tr class="odd">
+<td>static AppRestartFailureReason <strong>RequestRestartNow</strong>(string restartArguments)</td>
+<td><p>Equivalent to CoreApplication.RequestRestartAsync, except that it is <em>not</em> async.
+<ul>
+<li>If the restart fails, but the user subsequently launches the app manually, the app will launch normally and no restart arguments will be passed. </li>
+<li>If the app has any in-process background tasks running when it calls this API, those tasks will be cancelled in the normal way. Out-of-process background tasks will not be affected. </li>
+</ul>
+ The restarted instance will be activated with Windows.ApplicationModel.Activation.LaunchActivatedEventArgs.</p>
+ <ul>
+<li>When the app is restarted, LaunchActivatedEventArgs.PreviousExecutionState will have the value Terminated so that the app can distinguish between a resume and a restart. </li>
+</ul>
+<p>If the request to restart is successful, the application will be restarted immediately. The user does not have to do anything else/handle a return value in any way.
+
+If the request fails due to a nonfatal reason, a failure reason is returned for why the restart didn’t successfully complete. This method leverages an existing enum (AppRestartFailureReason) which has the following relevant return fields.
+<ul>
+<li> InvalidUser: Could not restart for the specified user.</li>
+<li> RestartPending: A restart is already in progress.</li>
+<li> Other: Unspecified failure.</li>
+</ul>
+</p></td>
+</tr>
+</tbody>
+</table>
+
+When an application calls RequestRestartNow, the mechanism for restart will be as follows:
+1. Application calls RequestRestartNow API.
+2. The API will CreateProcess with the Agent (a helper EXE in the framework package containing the API) and will then wait for an an acknowledgment (ACK) from the Agent.
+3. The Agent will get the executable path of the application and signal an ACK to the API. The Agent will then then terminate the application and then CreateProcess with the application (restarted).
```

Comment:
> The Agent will then terminate the application [...]

Am concerned about cleanup here, if the Agent handles app termination.

An app may only cleanup unmanaged resources (e.g. shared GDI handles) when terminating cleanly. Is the thinking here that the developer needs to redesign their application to have a cleanup-first-then-call-RequestRestartNow code path? Perhaps this can be redesigned so that the Agent sends back an ACK to the caller, whom can then attempt to terminate gracefully?

Reply:
Apps already have to deal with abnormal termination in the case of termination-on-suspension, app crash, or the user terminating from Task Manager. Resources like kernel handles and GDI resources are automatically cleaned up at process termination. Other resources like temporary files will have to be cleaned up by the restarted application, just like any other launch after termination. So no new recovery scenarios are introduced here.

Reply:
I was thinking more along the lines of shared GDI resources like icons and possible impact to Desktop Bridge middleware, but I concede this all falls under abnormal termination handling.